src/openrct2/peep/Guest.cpp

```diff
@@ -441,17 +441,23 @@ bool Guest::GuestHasValidXY() const
     return false;
 }
 
-void Guest::ApplyEasterEggToNearbyGuests(easter_egg_function easter_egg)
+template<void (Guest::*F)(Guest*)> static void ApplyEasterEggToNearbyGuests(Guest* guest)
 {
-    if (!GuestHasValidXY())
+    const auto guestLoc = guest->GetLocation();
+    if (!IsValidLocation(guestLoc))
         return;
 
-    for (auto* otherGuest : EntityTileList<Guest>({ x, y }))
+    for (auto* otherGuest : EntityTileList<Guest>(guestLoc))
     {
-        auto zDiff = std::abs(otherGuest->z - z);
+        if (otherGuest == guest)
+        {
+            // Can not apply effect on self.
+            continue;
+        }
+        auto zDiff = std::abs(otherGuest->z - guestLoc.z);
         if (zDiff <= 32)
         {
-            (*this.*easter_egg)(otherGuest);
+            (*guest.*F)(otherGuest);
```

Comment:
If I'm not mistaken*, with `std::invoke` you should be able to simplify that awkward notation and `F` could potentially become a `typename Pred`, much like in many std functions.

*I haven't tested that.

Reply:
I used a strict type so its clear from the beginning what this function expects. std::invoke is a good point.